sdk/core/src/policies/telemetry_policy.rs

```diff
@@ -0,0 +1,131 @@
+use std::env::consts::{ARCH, OS};
+use std::sync::Arc;
+
+#[derive(Clone, Debug, Default)]
+pub struct TelemetryOptions {
+    pub application_id: Option<String>,
+}
+
+#[derive(Clone, Debug)]
+pub struct TelemetryPolicy {
+    header: String,
+}
+
+/// Sets the User-Agent header with useful information in a typical format for Azure SDKs.
+///
+/// Client libraries should create a `TelemetryPolicy` using `option_env!()` like so:
+/// ```
+/// use azure_core::policies::{TelemetryOptions, TelemetryPolicy};
+/// let policy = TelemetryPolicy::new(option_env!("CARGO_PKG_NAME"), option_env!("CARGO_PKG_VERSION"), &TelemetryOptions::default());
+/// ```
+impl<'a> TelemetryPolicy {
+    pub fn new(
+        crate_name: Option<&'a str>,
+        crate_version: Option<&'a str>,
+        options: &TelemetryOptions,
+    ) -> Self {
+        Self::new_with_rustc_version(
+            crate_name,
+            crate_version,
+            option_env!("AZSDK_RUSTC_VERSION"),
+            options,
+        )
+    }
+
+    fn new_with_rustc_version(
+        crate_name: Option<&'a str>,
+        crate_version: Option<&'a str>,
+        rustc_version: Option<&'static str>,
+        options: &TelemetryOptions,
+    ) -> Self {
+        const UNKNOWN: &'static str = "unknown";
+        let mut crate_name = crate_name.unwrap_or(UNKNOWN);
+        let crate_version = crate_version.unwrap_or(UNKNOWN);
+        let rustc_version = rustc_version.unwrap_or(UNKNOWN);
+        let platform_info = format!("({}; {}; {})", rustc_version, OS, ARCH,);
+
+        if let Some(name) = crate_name.strip_prefix("azure_") {
+            crate_name = name;
+        }
+
+        let header = match &options.application_id {
+            Some(application_id) => format!(
+                "{} azsdk-rust-{}/{} {}",
+                application_id, crate_name, crate_version, platform_info
+            ),
+            None => format!(
+                "azsdk-rust-{}/{} {}",
+                crate_name, crate_version, platform_info
+            ),
+        };
+
+        TelemetryPolicy { header: header }
+    }
+}
+
+#[async_trait::async_trait]
+impl Policy for TelemetryPolicy {
+    async fn send(
+        &self,
+        ctx: &mut Context,
+        request: &mut Request,
+        next: &[Arc<dyn Policy>],
+    ) -> PolicyResult<Response> {
+        request
+            .headers_mut()
+            .insert(USER_AGENT, HeaderValue::from_str(&self.header).unwrap());
+
+        next[0].send(ctx, request, &next[1..]).await
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn test_without_application_id() {
+        let policy = TelemetryPolicy::new_with_rustc_version(
+            Some("azure_test"), // Tests that "azure_" is removed.
+            Some("1.2.3"),
+            Some("4.5.6"),
+            &TelemetryOptions::default(),
+        );
+        assert_eq!(
+            policy.header,
+            format!("azsdk-rust-test/1.2.3 (4.5.6; {}; {})", OS, ARCH)
```

Comment:
Don't most user-agents just do app/version component/version? #65 Do you have Azure examples or a spec on this format?

Reply:
All our SDKs includes runtime/compiler version, and something of the OS as well. This is our standard format. Since there's obviously no runtime, I'm using the compiler version here, if possible.